[api-extractor] Deeply linked types

rbuckton opened the pull request:
From https://github.com/microsoft/web-build-tools/pull/1317#issuecomment-500159936. This adds support for deeply linked types when emitting to Yaml for DocFX. It does this by utilizing DocFX's support for language-specific [type specifications in references](https://dotnet.github.io/docfx/spec/metadata_dotnet_spec.html#7-reference).

AlexJerabek commented:
Thanks for putting this PR together @rbuckton. I tested this on the [office-js doc set](https://github.com/OfficeDev/office-js-docs-reference) and received the following error:
```
Error: JSON validation failed:
yaml\excel\excel.bindingcollection.yml
Error: #/references
       Array items are not unique (indexes 0 and 1)
```
I'd love to sync up and continue testing this to ensure it works with our doc set and fixes #867.

rbuckton commented:
Can you reply with a copy of the relevant portion of the yaml file and I'll look into what is causing the duplication?

AlexJerabek commented:
[Here's a link to the JSON file that caused the error](https://raw.githubusercontent.com/OfficeDev/office-js-docs-reference/master/generate-docs/json/excel.api.json).

[Here's a gist with the JSON](https://gist.github.com/AlexJerabek/6977c5e3a7d957fb8332ce9b361ef55d) for the `Excel.BindingCollection` mentioned in the above error.

Please let me know if you need anything else.

rbuckton commented:
The `yaml\excel\excel.bindingcollection.yml` file mentioned in the error would be more useful for diagnosing the issue

rbuckton commented:
@AlexJerabek should be addressed by e9d3df9

AlexJerabek commented:
@rbuckton Tested the latest version on [this staging branch](https://review.docs.microsoft.com/en-us/javascript/api/excel/excel.chart?view=office-js&branch=AlexJ-WBTest). Looks good. Thank you.

 The only problem is the changes break our custom example snippet mapping in [OfficeYamlDocumenter](https://github.com/microsoft/web-build-tools/blob/master/apps/api-documenter/src/documenters/OfficeYamlDocumenter.ts). The `console.error(colors.yellow('Warning: Unused snippet ' + apiName));` is being hit for every snippet. That is probably on us to investigate and fix, but if you see an obvious naming change, please let us know.

AlexJerabek commented:
Looks like the issues in `OfficeYamlDocumenter` are because of the improved UIDs. That's on us to fix our mapping files. Thanks for addressing #867.

octogonz commented:
@rbuckton following up from Gitter:

My main feedback was that the UID class seems out of place in **api-extractor-model** and as a field in the .api.json file format, because the Universal ID concept is specific to DocFX.  This new notation solves the same problems as TSDoc "declaration references" (used in `{@link}` tags for example).   Your notation improves on the existing syntax by (1) increasing compatibility with JSDoc namepaths and (2) simplifying the strings overall.  Thus it makes sense to integrate this into TSDoc, rather than introducing a second different notation that is DocFX-specific.

Also, the code review may go quicker if the UID parser/emitter is a separate PR from the API Documenter updates (even though it's certainly helpful to see them together in https://github.com/microsoft/web-build-tools/commit/e9d3df9d177413d3a29777ecc025698c5aad83ea).

That said, you have fixed an important problem for YamlDocumenter that I'm sure @AlexJerabek is eager to use, so we should be careful not to delay things unnecessarily.  Thus I'd like to suggest the following:

1. Rename Uid.ts to DeclarationReference.ts
2. Open a separate PR to add this new DeclarationReference.ts class under a [tsdoc/src/beta](https://github.com/microsoft/tsdoc/tree/master/tsdoc/src) in the TSDoc project, which we can designate as a "proposed syntax change" not ready for general usage yet
3. Then update your PR here to import the "beta" class from **@<!---->microsoft/tsdoc** package.
4. Rename .api.json field from `"uid"` to `"declarationReference"` (or some other name that is not DocFX-specific). This can completely replace `"canonicalReference"`.  That field was halfbaked and not really used by the deserializer or API Documenter, so I wouldn't consider it a breaking API change to remove.

After both PRs are merged, the follow up steps would be:
- Create a PR updating [DeclarationReferences.ts](https://github.com/microsoft/tsdoc/blob/master/spec/code-snippets/DeclarationReferences.ts) to show how it would look with your new notation, and solicit feedback from some interested people
- If everyone is cool with it, then we can modify [NodeParser.ts](https://github.com/microsoft/tsdoc/blob/master/tsdoc/src/parser/NodeParser.ts) to integrate your parser/emitter into the `DocDeclarationReference` API.  This /will/ be a breaking change for both TSDoc and API Extractor, both API and input specification.  We would major bump API Extractor to version 8 for that change.
- Incorporate your notation into a formal TSDoc spec document (which we've been planning to write up for a while; the [ae-ambiguous-ids](https://github.com/microsoft/web-build-tools/issues/1308) problem you're helping with here is one of a few remaining loose ends that I wanted to solve before writing up the spec.)

(I'm not asking you to do these follow up steps, since your goal in this PR was merely to improve the doc generation. We can handle this part.  If you /do/ have time to help with that, it would be super awesome of course. :-) )

octogonz commented:
Also congrats on getting the 1337 integer value! 😎

rbuckton commented:
> Thus I'd like to suggest the following:

I will see if I can put together a PR for tsdoc next week.

octogonz commented:
> I will see if I can put together a PR for tsdoc next week.

@rbuckton Do you want some help with this?  I might have some time over the holiday.

AlexJerabek commented:
Hi @rbuckton. What's the status of this PR? We're eager to integrate these changes with our documentation tooling.

octogonz commented:
@AlexJerabek We're workin' on it.  I merged and published https://github.com/microsoft/tsdoc/pull/172 today, so this PR should be unblocked.

octogonz commented:
@rbuckton I'm trying to upgrade your PR branch to use the TSDoc version of the API, but I don't seem to have permissions to push to your branch.  Could you fix that?

octogonz commented:
> 4\. Rename .api.json field from `"uid"` to `"declarationReference"` (or some other name that is not DocFX-specific). This can completely replace `"canonicalReference"`.  That field was halfbaked and not really used by the deserializer or API Documenter, so I wouldn't consider it a breaking API change to remove.

I finished steps 1-3.  I've been thinking about the name for the .api.json field.

TSDoc's **declaration reference** is an expression that maps to a declaration.  To facilitate human authoring, there can be many equivalent references for the same declaration.  For example, these all might be equivalent and unambiguous (in a certain context):
- `my-package!someApi:function(1)`
- `my-package!someApi:function`
- `my-package!someApi`
- `my-package!"someApi"`
- `someApi`

DocFX does not understand this equivalence relation.  The DocFX **UID** (unique identifier) is a simple text string, and it is compared using a simple string comparison.  Each API declaration gets exactly one UID, and hyperlinks must use that exact string.

If we can define a function that produces `my-package!someApi:function(1)` from any of the above equivalent inputs, this is called [canonicalization](https://en.wikipedia.org/wiki/Canonicalization).  It picks out a "canonical" form.  This idea is slightly stronger than the DocFX UID, since it implies that any tool which examines APIs will agree on the canonical form.  But I think your design has that property. That's why I was calling it a **canonical reference**.

But when a thing has a unique identifier, it's weird to talk about it as being a "reference".  References are possible ways of mentioning about something, whereas the official label for that thing is its "name".  

Thus I'm thinking the terminology should be like this:

"A **declaration reference** is an expression used to refer to a TypeScript declaration."
"The **canonical name** is the unambiguous, unique, normalized form of a **declaration reference**."
"For TypeScript we use the TSDoc **canonical name** as the DocFX UID."

octogonz commented on apps/api-extractor-model/src/items/ApiDeclaredItem.ts:
```diff
@@ -42,13 +46,17 @@ export class ApiDeclaredItem extends ApiDocumentedItem {
 
     options.excerptTokens = jsonObject.excerptTokens;
+    if (jsonObject.uid !== undefined) {
+      options.uid = DeclarationReference.parse(jsonObject.uid);
```
@rbuckton Just curious, do we actually need to parse this?  Seems like the .api.json hierarchy already has all the information needed to reproduce this UID.  If so, we could think of this field as being serialized for informational purposes only.

Suppose the canonicalization rules are changed in the future.  If we deserialize the UID, we'll get the old canonical form, whereas if we generate it then we'd get the new canonical form.  I think the latter is more desirable, but I'm not 100% sure.

rbuckton replied:
I parse the UID for several reasons:
- Validation of the UID in the event some external tool performs preprocessing on the json.
- Handling a possible future case of the UID being customized by a doc comment.
- The UID was likely generated with full access to the TypeScript type checker. It's possible that some fidelity might be lost when constructing the UID merely from the object model.

That said, I don't have a strong preference.

octogonz replied:
Thanks for explaining. If you can think of any discrepancies between the model and the type checker, I'd be interested to understand then and see if we can incorporate them into the model.

octogonz replied:
BTW I realized that there is a subtlety to this "canonical" idea.  Suppose a package does this:

```ts
enum A { }
namespace A { }
export { A as B };
export { A as C };

// Let's say D is also an enum+namespace
export { D as E, D as F } from 'other-package';
```

How many "declarations" are there here to be canonicalized?  Today API Extractor's answer would be:

- `this-package!B:enum`
- `this-package!B:namespace`
- `this-package!C:enum`
- `this-package!C:namespace`
- `other-package!D:enum`
- `other-package!D:namespace`
- `this-package!E:import`  // modeled by AstImport today but not written to .api.json yet
- `this-package!F:import`

Unexported declarations like `A` are considered anonymous and the .d.ts rollups are free to fiddle with their names.  

And if an aliased symbol gets exported multiple times, each export is considered to be its own documentation object. `D` and `E` are aliases for the same symbol, but API Extractor sees them as not even being the same kind of thing.  This is justified because alias equivalence depends on the version of `other-package` that gets installed, which may be unknown (e.g. due to an NPM peer dependency).

The weirdest case is `B` and `C`, which today have all their members duplicated in the generated documentation.  Really one of them should be recast as `this-package!C:alias`, but without a special TSDoc tag, there's no deterministic way to choose.  (Fortunately this rarely arises in practice.)  

But even if we fixed that, I think I'd still consider `this-package!C:alias` to be "canonical" according to my definition, because when someone links that in the docs, they really are talking about `C` and not `B` or `A`.

I feel fairly content with API Extractor's simplifications, because philosophically I see TypeScript as two completely different languages:
- A rich type algebra that enables IntelliSense for every possible legacy JavaScript API
- The subset of Java-like "good parts" that's sufficient to build scalable modern code bases without everyone needing a PhD in type algebra, while ensuring ["one-- and preferably only one --obvious way"](https://www.python.org/dev/peps/pep-0020/) to write code

API Extractor is much more focused on the second one.  But I recognize that this "canonical name" idea may fall apart when applied to the compiler's world, where every file is a module, and namepaths can refer to unexported "inner types".  

Let me know if you have any thoughts about that.

octogonz replied:
Hrmm... perhaps this:
```ts
- `this-package!E:import`  // modeled by AstImport today but not written to .api.json yet
- `this-package!F:import`
```

Should have been emitted like this:
```ts
- `this-package!E:enum`  // modeled by AstImport today but not written to .api.json yet
- `this-package!E:namespace`
- `this-package!F:enum`
- `this-package!F:namespace`
```

...so people can refer to individual declarations of an imported symbol.  But now I'm thinking I should just call it a **canonical reference** instead of a **canonical name**, because the thing being named is more nebulous than I original thought.

octogonz replied:
Hmm... today these edge cases are absent from the .api.json.  For example [test-outputs/exportDuplicate](https://github.com/rbuckton/web-build-tools/blob/9d3f6fa22ee8c30f3ed7a10e8745319339818a39/build-tests/api-extractor-scenarios/etc/test-outputs/exportDuplicate/api-extractor-scenarios.api.json#L30) is missing most of the declarations from [api-extractor-scenarios.api.md](https://github.com/rbuckton/web-build-tools/blob/deeplyLinkedtypes/build-tests/api-extractor-scenarios/etc/test-outputs/exportDuplicate/api-extractor-scenarios.api.md).  We should get that fixed, since it's a super interesting case for your UID generation algorithm.

I opened https://github.com/microsoft/web-build-tools/issues/1404 to track this.

rbuckton replied:
I'm not sure the best way to handle this. From TypeScript's perspective, `B`, `C`, `E`, and `F` would all be an _alias_ for the real declaration. If it were me, I would have the canonical name for `B` and `C` be something like `this-package!B:alias` and `this-package!C:alias` respectively, with references to `this-package!A:enum`, etc. The problem being that DocFX doesn't understand this right now (and thus would require a custom template to support it).

The documentation might look something like:

---

## Enum `A` (local)
(id: `this-package!~A:enum` (local, non-exported))
(definition of `A` as an enum)

## Namespace `A` (local)
(id: `this-package!~A:namespace` (local, non-exported))
(definition of `A` as a namespace)

## Export `B`
(id: `this-package!B:alias`)
Export of:
- **Enum `A` (local)**
- **Namespace `A` (local)**

## Export `C`
(id: `this-package!C:alias`)
Export of:
- **Enum `A` (local)**
- **Namespace `A` (local)**

---

Alternatively, you define `B` and `C` per meaning:

---

## Enum `A` (local)
(id: `this-package!~A:enum` (local, non-exported))
(definition of `A` as an enum)

## Namespace `A` (local)
(id: `this-package!~A:namespace` (local, non-exported))
(definition of `A` as a namespace)

## Enum `B` (alias)
(id: `this-package!B:enum`)
Export for: **Enum `A` (local)**

## Namespace `B` (alias)
(id: `this-package!B:namespace`)
Export for: **Namespace `A` (local)**

## Enum `C` (alias)
(id: `this-package!C:enum`)
Export for: **Enum `A` (local)**

## Namespace `C` (alias)
(id: `this-package!C:namespace`)
Export for: **Namespace `A` (local)**

---

If you are doing .d.ts rollup, then you might just chose one of the two (`B` or `C`) to be the definition and just have the other one point to the first one.

For a re-export, you can either do the same as above or just use the IDs for the thing being exported.

rbuckton replied:
Another option would be to update DeclarationReference to support further disambiguation:

```
this-package!E:export(other-package!D:enum)
this-package!E:export(other-package!D:namespace)
```

Though that would obviously complicate the logic of resolving a partial ID like `E:enum` to the fully-qualified canonical version.

octogonz replied:
> If it were me, I would have the canonical name for `B` and `C` be something like `this-package!B:alias` and `this-package!C:alias` respectively, with references to `this-package!A:enum`, etc.

Here's a more extreme example:

**package-1**
```ts
export function A(): void {
}

export interface A {
}
```

**package-2**
```ts
export { A as B } from 'package-1';
```

**package-3**
```ts
import { B } from 'package-2';

export class C implements B {
}
```

Now suppose that we want to make the hyperlink for `C`'s `implements B` excerpt.  In order to take us to the right DocFX page (and ultimately be able to follow hyperlinks to the right final definition), that reference needs to tell us that:

1\. The target is called `B`.
2\. The target comes from `package-2`
3\. The target is the interface, not the function

All 3 pieces of information are required for an accurate hyperlink.  Thus I'd argue the canonical reference must be `package-2!B:class`.

Now, generally your UID also tells us what kind of DocFX page we'll find when we follow the hyperlink.  Here, that requires an extra piece of information:

4\. The target is actually an alias for some other declaration.

But is this extra information important?  What would we need it for?  For a **declaration reference** my goal was merely to ensure that it uniquely identifies the target `ApiItem`.

When I click on the hyperlink for `C`'s `implements B` excerpt, I'd expect to go to a page something like this:


> ---
> 
> ## Interface `B` (alias)
> (id: `package-2!B:class`)
>
> Alias for: <a href="#">A</a> from <a href="#">package-1</a>
> 
> **Signature**
> ```ts
> export { A as B } from 'package-1';
> ```
> ---

rbuckton replied:
I agree, its somewhat similar to showing inherited members in a subclass's documentation.

octogonz commented on apps/api-documenter/src/documenters/YamlDocumenter.ts:
```diff
@@ -554,33 +606,50 @@ export class YamlDocumenter {
+      return apiItem.uid;
+    }
 
+    let result: DeclarationReference = new DeclarationReference();
```
While debugging, I am finding that `UidGenerator._symbolToUid()` returns `undefined` for many of the package exports.  A ton of .api.json entries are missing the `uid` field.  Some of it is just bugs that can be fixed.  But fundamentally `UidGenerator._symbolToUid()` walks backwards from the symbol, trying to reach a package entry point, and returning `undefined` if not.  For an exported symbol, I find it difficult to prove that this will always produce a UID.  And even if it does, the UID may disagree with the path that API Extractor follows to reach the `AstDeclaration`.  This disagreement would occur because API Extractor's analysis crawls forward from the entry point, and thus aliases may be traversed along a different path.

The code here in `_getUidObject()` looks like it compensates for the missing UIDs by constructing them from the .api.json model.  But this seems hacky; every item in the .api.json should have a UID.

Thus I'm thinking it might be safer to do this:

1. For exported APIs, always calculate their UID based on API Extractor's model
2. Do it in **api-extractor-model** instead of here in **api-documenter**
3. For excerpt links, first check whether we can match the symbol to something that was crawled by the `Collector` and already has a known UID; this is the idea that @adventure-yunfei used in https://github.com/microsoft/web-build-tools/pull/1005
4. Otherwise, we could use `UidGenerator._symbolToUid()` as a heuristic to make a best attempt.  However I also notice that `_symbolToUid()` is going to produce a lot of UIDs that don't correspond to any DocFX page.  I wonder if that is going to have a nontrivial performance cost.

@rbuckton Unless you have any objections, I'm going to take a shot at doing (1) and (2) as part of this PR, since seeing the correct UID for every crawled item is important for test validation of our spec.  Whereas (3) and (4) just need to be roughly correct; if some hyperlinks are missing, those are minor bugs that we can fix over time.

rbuckton replied:
The main reason to construct a UID from a Symbol is to get an accurate canonical ID for a reference to a declaration/type that is not included in your package's .api.json (such as from an external package). This is so that if the documentation for the two packages were to be merged, or if the external package provided an *xrefmap.yml* file for its documentation, that the UIDs would align and the documentation links would be seamless.

octogonz replied:
@rbuckton That makes sense.  However for the APIs that /are/ part of the .api.json, it's much simpler and more reliable to calculate them from the model.  So far I haven't encountered any cases where information is lost compared to the other approach.  (Well, ECMAScript symbols are not handled correctly, but that's merely an incomplete feature that should be solved by improving **api-extractor-model**.)

I peeled this work off into a separate [PR #1406](https://github.com/microsoft/web-build-tools/pull/1406).  Could you take a look at it and let me know if it looks right?  I am mostly interested in whether I used your class correctly, and whether the generated UIDs match your design.  I'll add some specific questions as comments to that PR.

octogonz commented:
For people following this PR, I posted a status update in https://github.com/microsoft/web-build-tools/issues/1252#issuecomment-516258326:

> Here's where we stand:
> 
> * [PR 1406](https://github.com/microsoft/web-build-tools/pull/1406) introduces a `canonicalReference` field in .api.json that correctly distinguishes static/instance members (as well as all the other cases such as function overloads).  This is the difficult, important piece of technology that is the basis for solving all the [ae-ambiguous-ids](https://github.com/Microsoft/web-build-tools/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Aae-ambiguous-ids) issues
> * But PR 1406 is blocked by [TSDoc PR 174](https://github.com/microsoft/tsdoc/issues/174) due to a bug in the new parser.  This bug causes API Extractor to fail when processing `@microsoft/rush-stack-compiler-3.5`, so it is a serious issue that needs to be fixed.  The fix requires a grammar change, so it's not super hard but not super easy either. Either @rbuckton or I will fix it as soon as we get a chance.
> * Once those two PRs are merged, the solution to your fix is essentially already present in the [PR 1337](https://github.com/microsoft/web-build-tools/pull/1337) branch.  But that branch is fairly big, so we've been peeling off pieces into smaller PRs, rather than merging it directly.  If you need a quick fix for #1252, we could peel that off fairly easily.
> 
> So, in a nutshell, fixing [TSDoc PR 174](https://github.com/microsoft/tsdoc/issues/174) is what will get things moving again. We need to prioritize that.

AlexJerabek commented:
Hi @rbuckton and @octogonz, 
Judging by the linked issues in the previous post, this PR seems unblocked (which is super exciting). What additional efforts are needed to support the deeply linked types?

adventure-yunfei commented:
Looking forward to this PR too. And it would be better if markdown support is also on the way. 😃

rbuckton commented:
I've done some cleanup, rebased against master and squashed the disparate changes to clean up the PR a bit.

Some notes:
- "Reference" excerpt tokens now have a `canonicalReference` added when the reference is not for a declaration name (i.e. skips adding references for parameter names, etc.).
- The `YamlDocumenter` uses the `canonicalReference` to build the complex type for display purposes.

I may need to clean up the `_renderType` method to only do anything complex when there is at least *one* ExcerptToken with a `canonicalReference`, otherwise there's no point in writing out a yaml reference. If we assume the api.json file we're using is the source of truth (and that valid `canonicalReference` entries are added for each ExcerptToken), then there's no sense in trying to derive a `uid` when we can't find one.

That means we can get rid of things like `_knownTypeParameters`, `_linkToUidIfPossible`, and `_apiItemsByTypeName`, since we should assume that the `canonicalReference` of the ExcerptToken is correct (and if it isn't it's a bug in `DeclarationReferenceGenerator` in `api-extractor`).

octogonz commented:
Awesome, thanks very much!  I'll hopefully have some time to review this over the weekend.  (Been busy recently trying to wrap up the migration from TSLint -> ESLint+TypeScript.)

rbuckton commented:
@octogonz Can you take another look?

octogonz commented on build-tests/api-extractor-scenarios/etc/test-outputs/typeOf/api-extractor-scenarios.api.json:
```diff
@@ -81,7 +82,8 @@
               "kind": "Reference",
-              "text": "ForgottenExport"
+              "text": "ForgottenExport",
+              "canonicalReference": "!ForgottenExport:class"
```
`ForgottenExport` is linked as a global (ambient) declaration.  But it is effectively an unreachable declaration.  It would be more correct to omit the `canonicalReference` entirely.  (It could also be modeled using JSDoc's `~` "inner member" notation.  But currently API Extractor has the philosophy that if something is important enough to be documented, it should be importable.)

rbuckton replied:
That philosophy may not match with TypeScript output as we may sometimes manufacture non-exported types in declaration files as a result of inference.

rbuckton replied:
I've improved the declaration reference generation for local (non-exported) types.

octogonz replied:
> we may sometimes manufacture non-exported types in declaration files as a result of inference.

Are you referring to [this sort of thing](https://github.com/microsoft/web-build-tools/blob/16c6a7e36f5d67c4aae8dda17b7052e6ddb3ce4e/apps/api-extractor-model/src/model/ApiConstructSignature.ts#L66), where `extends ApiTypeParameterListMixin(ApiParameterListMixin(ApiReleaseTagMixin(
  ApiReturnTypeMixin(ApiDeclaredItem))))` gets separated into `ApiConstructor_base`?

**lib/model/ApiConstructor.d.ts**
```ts
declare const ApiConstructor_base: typeof ApiDeclaredItem & (new (...args: any[]) => ApiReleaseTagMixin) & (new (...args: any[]) => ApiParameterListMixin);

/**
 * Represents a TypeScript class constructor declaration that belongs to an `ApiClass`.
 * @public
 */
export declare class ApiConstructor extends ApiConstructor_base {
    static getContainerKey(overloadIndex: number): string;
    constructor(options: IApiConstructorOptions);
    /** @override */
    readonly kind: ApiItemKind;
    /** @override */
    readonly containerKey: string;
    /** @beta @override */
    buildCanonicalReference(): DeclarationReference;
}
```

rbuckton replied:
That is one case, though I think there are others as well.

octogonz commented on build-tests/api-documenter-test/etc/api-documenter-test.api.json:
```diff
@@ -1471,6 +1475,157 @@
+                {
+                  "kind": "Reference",
+                  "text": "T",
+                  "canonicalReference": "T"
```
This corresponds to `genericReferenceMethod<T>(x: T): T;` from `IDocInterface6`.

Is it intentional to generate a reference for a type parameter?  That seems not useful.

rbuckton replied:
I can remove the canonical reference.

octogonz commented on apps/api-documenter/src/documenters/YamlDocumenter.ts:
```diff
@@ -105,67 +118,94 @@ export class YamlDocumenter {
+    const savedKnownTypeParameters: Set<string> | undefined = this._knownTypeParameters;
+    try {
+      // Track type parameters declared by a declaration so that we do not resolve them
+      // when looking up types in _linkToUidIfPossible()
```
In the test, this input:
```ts
export interface IDocInterface6 {
  . . .
  genericReferenceMethod<T>(x: T): T;
}
```

...gets emitted as this YAML:
```yaml
  - uid: 'api-documenter-test!IDocInterface6#genericReferenceMethod:member(1)'
    name: genericReferenceMethod(x)
    fullName: genericReferenceMethod(x)
    langs:
      - typeScript
    type: method
    syntax:
      content: 'genericReferenceMethod<T>(x: T): T;'
      return:
        type:
          - T
        description: ''
      parameters:
        - id: x
          description: ''
          type:
            - T
      typeParameters:
        - id: T
```
...with a reference like this:
```yaml
references:
  - uid: T
```

If we really want to treat `T` as a referencable thing, then it probably should get a unique UID notation (e.g. something like `api-documenter-test!IDocInterface6#genericReferenceMethod:member(1).T`).  The `this._knownTypeParameters` lookup here appears to be a workaround for the problem that the name `T` is used for many different things, but is being treated as a "declaration reference".

But stepping back for a moment, I'm not sure I see any value in linking to generic parameter declarations, for the same reason that we don't link to regular parameters either.  They are very local things, and their documentation tends to be very small, so it's quite sufficient to link to the associated function/method/class instead.

@rbuckton Thus, it seems like we could completely eliminate `this._knownTypeParameters` logic by not generating a `canonicalReference` for `T` in the first place.

rbuckton replied:
The reason this is here is to prevent a possible collision with a type in scope that has the same name as a type parameter:

```ts
interface Foo {}
declare function f<Foo>(f: Foo): Foo;
```
In the output yaml, the `Foo` interface has `uid: !Foo:interface` and `name: Foo`. Unfortunately, I cannot recall whether docfx will attempt to resolve a `type: Foo` against a local `name:` if it can't resolve the uid, but I believe that was the reason why I added this in the first place. It's been more than a few weeks since I first looked at this.

rbuckton replied:
I can remove this logic. We can decide if we want to add it back in if we see issues with docfx.

octogonz replied:
If I remember right, DocFX's file format has a design flaw where if a `type` value matches a UID, then it is assumed to be a reference to that UID.  They provide no escaping mechanism that could be used to represent a token that accidentally matches a UID, but is meant to be plain text (because it does not correspond to any linkable object).  If that's true, it is a flaw in the file format.

However, luckily your "canonical reference" notation will always contain a `!` character, which never appears in a valid type expression.  Thus, API Extractor should never encounter such a conflict.

It was possible type parameters only because the generic parameter is appearing in the `canonicalReference` field:

```js
{
  "kind": "Reference",
  "text": "T",
  "canonicalReference": "T"  // <=== not a correct reference
},
```

octogonz commented on apps/api-extractor/src/generators/ExcerptBuilder.ts:
```diff
@@ -113,8 +118,12 @@ export class ExcerptBuilder {
+        const canonicalReference: DeclarationReference | undefined = isDeclarationName(name)
+          ? undefined
+          : state.referenceGenerator.getDeclarationReferenceForIdentifier(name);
         ExcerptBuilder._appendToken(excerptTokens, ExcerptTokenKind.Reference,
```
BTW for an input like this:
```ts
export function exampleFunction(x: number): number {
  return x;
}
```

...it was an accident that identifiers like `exampleFunction` and `x` are being represented as `ExcerptTokenKind.Reference`.  The initial implementation merely sought to demonstrate the `Content`/`Reference` distinction for consumers of the API, but without actually emitting any link targets.  It simply assigned `ExcerptTokenKind.Reference`  whenever `span.kind === ts.SyntaxKind.Identifier`, without considering the kind of identifier.

Instead of this:
```js
          "excerptTokens": [
            {
              "kind": "Content",
              "text": "export declare function "
            },
            {
              "kind": "Reference",
              "text": "exampleFunction"
            },
            {
              "kind": "Content",
              "text": "("
            },
            {
              "kind": "Reference",
              "text": "x"
            },
            {
              "kind": "Content",
              "text": ": "
            },
            {
              "kind": "Content",
              "text": "number"
            },
            {
              "kind": "Content",
              "text": "): "
            },
            {
              "kind": "Content",
              "text": "number"
            },
            {
              "kind": "Content",
              "text": ";"
            }
          ],
```
...the `"kind"` should really be `"Content"` everywhere.  We should only use `"Reference"` when there we have an actual `"canonicalReference"` link target.

(BTW if it turns out to be useful to distinguish `"exampleFunction"`  versus `"): ", then I'd suggest to introduce a new kind like `ExcerptTokenKind.SimpleIdentifier` for identifiers that lack a `canonicalReference`.)  We could also maybe convert `IExcerptToken` into a discriminated union based on `ExcerptTokenKind`, to make that more clear.

rbuckton replied:
The problem I see is that the `ExcerptBuilder` doesn't know anything about the token in question. If we remove the `canonicalReference` for `exampleFunction` here, then we would have nothing to link to if we built an excerpt for the type `typeof exampleFunction`. It may be extraneous, but I don't think we should remove it.

octogonz replied:
Take a look at PR https://github.com/rbuckton/web-build-tools/pull/2 (into your PR branch).

I changed this:

[src/generators/ExcerptBuilder.ts ](https://github.com/rbuckton/web-build-tools/pull/2/commits/a2e45837f7971fef351acf943afebab77b48b3f6)
```ts
      if (span.kind === ts.SyntaxKind.Identifier) {
        const name: ts.Identifier = span.node as ts.Identifier;
        const canonicalReference: DeclarationReference | undefined = isDeclarationName(name)
          ? undefined
          : state.referenceGenerator.getDeclarationReferenceForIdentifier(name);
        ExcerptBuilder._appendToken(excerptTokens, ExcerptTokenKind.Reference,
          span.prefix, state, canonicalReference);
      } else {
        ExcerptBuilder._appendToken(excerptTokens, ExcerptTokenKind.Content,
          span.prefix, state);
      }
```
...to this:
```ts
      let canonicalReference: DeclarationReference | undefined = undefined;

      if (span.kind === ts.SyntaxKind.Identifier) {
        const name: ts.Identifier = span.node as ts.Identifier;
        if (!isDeclarationName(name)) {
          canonicalReference = state.referenceGenerator.getDeclarationReferenceForIdentifier(name);
        }
      }

      if (canonicalReference) {
        ExcerptBuilder._appendToken(excerptTokens, ExcerptTokenKind.Reference,
          span.prefix, state, canonicalReference);
      } else {
        ExcerptBuilder._appendToken(excerptTokens, ExcerptTokenKind.Content,
          span.prefix, state);
      }
```

The [test-outputs/typeOf2](https://github.com/rbuckton/web-build-tools/blob/585184975a21435ca9a88410f7b03b05e64ab097/build-tests/api-extractor-scenarios/etc/test-outputs/typeOf2/api-extractor-scenarios.api.json#L58) still correctly links `typeof f`:

```ts
            {
              "kind": "Content",
              "text": "typeof "
            },
            {
              "kind": "Reference",
              "text": "f",
              "canonicalReference": "api-extractor-scenarios!f:function"
            },
```

octogonz commented on common/reviews/api/api-extractor-model.api.md:
```diff
@@ -613,12 +613,18 @@ export class Excerpt {
+    // (undocumented)
+    readonly canonicalReference: DeclarationReference | undefined;
+    // (undocumented)
+    static fromJSON(object: IExcerptToken): ExcerptToken;
```
Similar to [.NET](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions), we use `camelCase` for acronyms with more than 2 letters.  So the name would be `fromJson` instead of `fromJSON`.

rbuckton replied:
However, ECMAScript specifically defines `toJSON` with upper-case for `JSON.parse`/`JSON.stringify`. Having a `toJSON` and a `fromJson` seems confusing. I added them for convenience, but I could always inline them into their only usages in `ApiDeclaredItem`.

rbuckton replied:
I've just removed the members and inlined them into their use sites.

octogonz replied:
Cool 👍

octogonz commented on build-tests/api-extractor-scenarios/etc/test-outputs/ambientNameConflict/api-extractor-scenarios.api.json:
```diff
@@ -62,7 +63,8 @@
               "kind": "Reference",
-              "text": "MyPromise"
+              "text": "MyPromise",
+              "canonicalReference": "api-extractor-scenarios!Promise:class"
```
*Issue: No reference should be generated because the `api-extractor-scenarios` package does not export a symbol `Promise`.*

rbuckton replied:
The problem is that it imports `Promise` from `./localFile` as the alias `MyPromise`. I get the name `Promise` by resolving the alias.

Perhaps this should have been `"api-extractor-scenarios/localFile!Promise:class"`? The problem with that is that api-extractor doesn't currently seem to support anything other than package-level exports.

Also, you include the definition of `MyPromise` as an unexported class in the rollup.d.ts named `Promise_2`, so perhaps this should have been `"api-extractor-scenarios!~Promise_2:class"`?

rbuckton replied:
Also, the `ExcerptBuilder` can't know whether the custom `Promise` here might be publicly exported as some completely different name by the package.

octogonz replied:
@rbuckton In @adventure-yunfei's PR https://github.com/microsoft/web-build-tools/pull/1005 (an earlier attempt at this), the approach was to build a [`Map<ts.Symbol, ApiItem>`](https://github.com/microsoft/web-build-tools/blob/f53c62accc4b0154a3b2a3ab00aa3a6ca7a8e812/apps/api-extractor/src/generators/ApiModelGenerator.ts#L48) for all exported APIs, and then consult that when generating the excerpt references.   (He didn't consider how to choose among multiple exports, but a heuristic would be reasonable.)

> Also, you include the definition of `MyPromise` as an unexported class in the rollup.d.ts named `Promise_2`, so perhaps this should have been `"api-extractor-scenarios!~Promise_2:class"`?

As far as I can see, the purpose of these references is for linking to documentation entries.  As long as API Extractor does not generate documentation for unexported symbols, then there isn't much benefit in referring to them.  The `api-extractor-scenarios.api.json` captures the complete inventory of possible documentation entries, we based on that we should be able to determine that `api-extractor-scenarios!Promise:class` isn't linkable, and thus should be omitted.

I'd like to get your PR merged though, so maybe we should postpone this issue and tackle it in a subsequent PR (unless it's a very simple addition).

rbuckton replied:
> The api-extractor-scenarios.api.json captures the complete inventory of possible documentation entries

I've mentioned before that this is an incorrect assumption. DocFX supports something called an `xrefmap.yml` file that lets you link to documentation *not* included in your project. This is useful if you are building an API on top of an external API and you want to link to the external API's documentation as well.

octogonz replied:
To clarify, I was referring to the subset of links that refer to declarations defined by the current package. For example, the `ts.SourceFile` containing the (followed) declaration for `MyPromise` is part of the file set for the current package, thus we can be sure there is no documentation entry for it.  There's no way it could appear in `xrefmap.yml`, right? 

For references to symbols imported from other packages, we should preserve those in case they are documented as you say. BTW if an external package publishes a [tsdoc-metadata.json](https://github.com/microsoft/tsdoc/issues/7#issuecomment-442271815) file, then we could possibly prove that its declarations would not appear in `xrefmap.yml` also, I think.

Does that make sense?

rbuckton replied:
> To clarify, I was referring to the subset of links that refer to declarations defined by the current package. 

What is the action item here then? Do I need to make a change in api-extractor to skip adding a `canonicalReference` to an excerpt for a non-exported member in the same package? Its unfortunate that those types will be printed in the excerpt but unreachable in the documentation. Should I be introducing a symbol -> ApiItem mapping as you mentioned to verify this, or some other mechanism?

> BTW if an external package publishes a tsdoc-metadata.json file, then we could possibly prove that its declarations would not appear in `xrefmap.yml` also, I think.

I suppose it does, though I'm not sure I want to take on the added complexity of that in this PR.

octogonz replied:
@rbuckton I think so, but let's postpone that for a later PR. I don't see any practical problems with the current implementation. As mentioned on Gitter, tomorrow we'll ask SharePoint to validate against their full data set, and if there are no issues then we should be good to finally merge it. (Hurray!)

BTW I'm really excited about this feature! Thanks so much for finishing it. After the PR is merged, I want to get it enabled for markdown output as well.

octogonz commented:
BTW here's an interesting edge case to think about (although possibly we can tackle it later in a separate PR):

```ts
// Let's say this is how A is named in the local file where it is declared
class A {
  public static create(): A {
    return new A();
  }
}

// But when we export from index.ts, we rename it to "NewA"
export { A as NewA }
```

The .d.ts rollup will look like this:
```ts
export declare class NewA {
    static create(): NewA;
}

export { }
```
...but the .api.json for `NewA.create()` looks like this:
```js
{
  "kind": "Method",
  "canonicalReference": "api-documenter-test!NewA.create:member(1)",  // <== correct
  "docComment": "",
  "excerptTokens": [
    {
      "kind": "Content",
      "text": "static "
    },
    {
      "kind": "Reference",
      "text": "create"
    },
    {
      "kind": "Content",
      "text": "(): "
    },
    {
      "kind": "Reference",
      "text": "A",
      "canonicalReference": "!A:class" // <== incorrect
    },
    {
      "kind": "Content",
      "text": ";"
    }
  ],
  "isStatic": true,
  "returnTypeTokenRange": {
    "startIndex": 3,
    "endIndex": 4
  },
  "releaseTag": "Public",
  "overloadIndex": 1,
  "parameters": [],
  "name": "create"
}
```

Ideally the return type should be rewritten as:
```ts
    {
      "kind": "Reference",
      "text": "NewA",
      "canonicalReference": "api-documenter-test!NewA.create:member(1)"
    },
```

rbuckton commented:
> [...] Ideally the return type should be rewritten as: [...]

What about cases like this?

```ts
class A {}
export { A as B, A as C };
```
What UID should use use for `B` and `C`? Do they end up with documentation entries that are just duplicates of `A`?

I'd rather the id for `A` be `my-package!~A:class`, since it is truly the canonical name. If you reference `B` or `C` as a type in your code, they should both link to `A`.

octogonz commented:
> What about cases like this?
> 
> ```ts
> class A {}
> export { A as B, A as C };
> ```
> 
> What UID should use use for `B` and `C`? Do they end up with documentation entries that are just duplicates of `A`?

The .d.ts rollup currently handles it that way:  Since `A` is not one of the exported names, then it's uncertain whether `B` or `C` should be the "real" name, so `A` is preserved locally in the rollup.

For the API documentation, it's not handled properly.  (I don't remember the details, but it's something wrong like no class at all, or only one entry but with the wrong name `A` instead of `B` or `C`.)  It's an important bug we should fix, but it hasn't been prioritized because it's somewhat rare.  

The most common situation is when an API got renamed, but they also want to preserve the old name to avoid breaking compatibility.

> I'd rather the id for `A` be `my-package!~A:class`, since it is truly the canonical name. If you reference `B` or `C` as a type in your code, they should both link to `A`.

This is theoretically the best answer, but I'm not sure it corresponds well with how the language is used in real life.  On the one hand, TypeScript is a general algebra that must describe every crazy thing you can do in untyped JavaScript.  But on the other hand, when people design APIs and want to publish professional docs, it needs to be simple and understandable.  

The story really should be: *"This class is officially called B.  But it has a secondary name C."*

We could equivalently say:  *"The class A has an unreachable name.  You must instead call it B or C when you import it.  In the docs, we'll randomly use both names for no good reason.  We won't call it A because some unrelated thing is exported with that name."*  This is theoretically identical, but it's a really terrible way to talk about your API.  :-)

Thus, it might be better to find some way to designate an "official" name, and then rewrite the signatures to banish `A` from the .api.json file and .d.ts rollup.

octogonz commented:
@rbuckton following up, yesterday we encountered a minor issue during testing.  I emailed you the repro.

AlexJerabek commented:
Thanks @rbuckton and @octogonz. We really appreciate this effort.
I'm working through the Office-JS tooling pipeline to get the code snippets pointing at the new UIDs. Are there any obvious changes you can see to make in [OfficeYamlDocumenter](https://github.com/microsoft/web-build-tools/blob/master/apps/api-documenter/src/documenters/OfficeYamlDocumenter.ts)?

octogonz commented:
FYI this was published as `api-documenter@7.4.1` and `api-extractor@7.4.0`.

> Are there any obvious changes you can see to make in [OfficeYamlDocumenter](https://github.com/microsoft/web-build-tools/blob/master/apps/api-documenter/src/documenters/OfficeYamlDocumenter.ts)?

Looking at the code, it seems like it should mostly work the same as before.  There are a few places that match `yamlItem.uid` that might possibly need adjustments.  

The way we validated SPFx was to generate docs in one folder using the old api-extractor and api-documenter (prior to Ron's change), and then in second folder using the new release.  Then we used [WinMerge](https://winmerge.org/?lang=en) to diff these two folders, and inspect how the .yml files had changed.

octogonz commented:
(@AlexJerabek Unrelated, I recently started introducing a plugin-in model for **api-documenter**, which will eventually allow the OfficeYamlDocumenter customizations to be separated into their own package, so they can be decoupled from the main **api-documenter** code base.  The more rigorous plugin contract should also reduce the likelihood of getting broken by code changes.  If you're curious, there's an example plugin [here](https://github.com/microsoft/web-build-tools/tree/master/repo-scripts/doc-plugin-rush-stack), which gets invoked by [this script](https://github.com/microsoft/web-build-tools/blob/master/repo-scripts/generate-api-docs/generate-api-docs.cmd) which uses [this config](https://github.com/microsoft/web-build-tools/blob/1fb8809098f3c284bbe1479b5e7d7135a64db35e/repo-scripts/generate-api-docs/api-documenter.json#L6).  I'll follow up with you later when this feature is more mature.)
